TEPHRA-96 Transaction checkpoints: support multiple write pointers per tx

ghelmling opened the pull request:
This change adds support for transaction checkpoints to Tephra.  Performing a transaction checkpoint will persist any pending writes on the client, and obtain a new write pointer for the transaction.  This allows clients to ensure that all writes prior to the checkpoint are visible for reads, while excluding the current write pointer from read operations where necessary.

A summary of the changes:
* separates the writePointer (same through the life of the transaction, used to identify the tx in TransactionManager), from the currentWritePointer (used to set the cell timestamp for writes)
* adds support for checkpoint write pointers in Transaction and to TransactionManager.InProgressTx
* adds a SnapshotCodecV4 to encode the additional InProgressTx properties
* updates TransactionVisibilityFilter to:
  * include writes from prior checkpoint write pointers in the tx
  * optionally exclude the current write pointer if TxConstants.TX_EXCLUDE_CURRENT_WRITE is set
* adds TransactionContext.checkpoint() along with RPC support to TransactionSystemClient and TransactionManager
* adds test for:
  * normal tx checkpoint visiblity, with and without TxConstants.TX_EXCLUDE_CURRENT_WRITE set
  * abort of checkpointed tx
  * invalidate of checkpointed tx

jwang47 commented on tephra-api/src/main/java/co/cask/tephra/Transaction.java:
```diff
@@ -48,35 +63,69 @@
+   * from {@link #getWritePointer()}.  This method should always be used when setting the timestamp for writes
+   * in order to ensure that the correct value is used.
+   */
+  public long getCurrentWritePointer() {
```
Should we have `getInitialWritePointer` and `getWritePointer` instead?

ghelmling replied:
I started down that path with TEPHRA-98 -- using "transaction ID" instead of "initial write pointer" -- but it had a number of drawbacks.  It's a very invasive change, touching many places just so we don't have to change the timestamp setting use of getWritePointer().  It's also backwards incompatible, which means it would likely be a very invasive change for deep users of Tephra like CDAP.  It's also ultimately not necessary to support this feature.  I found that it was much more straightforward and easier to understand if we simply changed the timestamp-setting usage of the writePointer.  From that standpoint, "currentWritePointer" was the best I could come up with.

anew replied:
I don't see how it is backward incompatible to use transactionID and writePointer. For any transaction client that existed before checkpoint, the transaction id and the write pointer will always be the same (because they don't enable checkpointing). 

I think currentWritePointer can be confusing, too, if we ever extend to allow a single transaction to have multiple active write pointers at the same time. (like we would for MapReduce if we assign a different write pointer to each mapper and reducer). 

Semantically (and from the description of this PR), it is more natural to call the "writePointer (same through the life of the transaction, used to identify the tx in TransactionManager)" the transactionId. Whereas currentWritePointer should be named writePointer. That is backward-compatible because existing clients call getWritePointer() to obtain the version with which to write their changes.

ghelmling replied:
s/getWritePointer()/getTransactionId()/ is also backwards incompatible for any clients calling it.  It's an API change that any clients currently using the writePointer to directly call TransactionSystemClient methods will have to make.

I agree that transactionId / writePointer naming is clearer, that's why I filed TEPHRA-98.  And since either change will introduce some incompatibility, I'll go ahead with it.

anew commented on tephra-api/src/main/java/co/cask/tephra/Transaction.java:
```diff
@@ -48,35 +63,69 @@
+   * Returns the initial write pointer assigned to the transaction.  This will remain the same for the life of the
+   * transaction, and uniquely identifies it with the transaction service.  This value should be provided
+   * to identify the transaction when calling any transaction lifecycle methods on the transaction service.
+   */
```
This is now not the write version any more, but it serves to identify the transaction. Therefore I would suggest to name it the transaction ID and call this method getTransactionId.

ghelmling replied:
The point in not renaming this was to minimize the changes needed by other clients.  However, since we had previously overloaded the use of this method (identifying the transaction for lifecycle methods, setting the write version), some usage will have to change.  Now that I've done the work splitting this out in Tephra, a rename should be straightforward.

abaranau replied:
+1 on rename

anew commented on tephra-api/src/main/java/co/cask/tephra/Transaction.java:
```diff
@@ -108,18 +157,60 @@ public long getFirstShortInProgress() {
+   * @param excludeCurrentWritePointer whether writes from the current write pointer should be visible
+   * @return true if the version is visible, false if it should be hidden (filtered)
+   */
+  public boolean isVisible(long version, boolean excludeCurrentWritePointer) {
```
I don't understand this API. Isn't it a property of the transaction itself, whether it includes the current write version or not? If there are two different APIs for this, then every transaction-aware must now know whether it is running inside of a "traditional" transaction or inside a check-pointed transaction. That means all existing transaction-awares must be changed to have that knowledge. It is much natural if the transaction-awares rely on the transaction to implement that knowldege.

anew replied:
See also my comment on the JIRA regarding this design point.

ghelmling replied:
Not sure I understand this.  The flag simply allows clients to exclude the current write pointer on a per-operation basis.  There is no need to know whether the transaction has been checkpointed or not.

From what I understand, you are suggesting that:
a) when first started, a transaction is obtained with read-your-writes behavior
b) after performing a checkpoint, read-your-write will no longer apply to the current write pointer.  To read the current write pointer, you would always have to perform a new checkpoint.

Is this correct?

I actually thought that we had worked out this discussion on the JIRA.  My plan for this change was:
a) by default, read-your-write semantics apply
b) the current write pointer can be excluded from reads by setting a flag on operations
c) used together this checkpoints, this allows clients (Phoenix) to read all prior writes in the transaction -- prior to a UPSERT ... SELECT .., perform a checkpoint, then set the exclude attribute on the Scan operation that is run.

That preserves the current read-your-writes behavior, while providing Phoenix with the tools it needs to exclude the current write pointer when necessary.

JamesRTaylor replied:
Thanks for the explanation, Gary. Yes, this will allow standard SQL semantics (i.e. a statement does not see its own writes) for relational database layers such as Phoenix.

abaranau replied:
> Isn't it a property of the transaction itself, whether it includes the current write version or not?

No. It is a property of an operation. I think it is good here

anew replied:
I am glad we all agree that we are mimicking SQL semantics. I relational databases, the visibility constraints or isolation level is typically configured for the life time of a transaction. I don't know of a way way in SQL to change the isolation level midway into a transaction (and if there is one, that is not a common pattern). So why would we want to set that for each operation? 

One advantage of making it a property of  the transaction is that semantics of a transaction are clean and straight-forward.  The other advantage is that we do not need to add a flag to every operation of every participating transaction-aware dataset, in order to participate in a check-pointed transaction.

Actually, coming back to @JamesRTaylor's remark on the Kira, in Phoenix, statements to not see their own writes. The way I understood that is that Phoenix will checkpoint the tx after every statement. My understanding was also that this applies to all statements in a transaction, including the first statement. @ghelmling I am not sure why you think that a transaction should start with "read-own-writes" and switch to "exclude-own-writes" at the time of the first Select-Upsert statement. If the first statement is a Select-Upsert, it should also not see its own writes, or should it?

ghelmling replied:
@anew You commented on the JIRA:

    An alternative implementation would be to have an option to exclude the current write version
    from a scan. That would simply add the current write version to the exclude filter before 
    submitting the scan request. All other operations would still "read own writes".

As far as I can see, this is exactly what this implementation provides.  You now seem to be arguing for something else.

You also stated:

    What I dislike about the proposed design is that it removes the "read your own writes" 
    capability before and after the checkpoint. Reads that happen after the checkpoint see writes 
    that happened before. But reads before checkpoints don't see prior writes, and reads after 
    checkpoint don't see other writes made after the checkpoint.

    That seems inconsistent with Tephra's normal snapshot isolation.

But this seems to be exactly what you are now proposing.

The goal of this issue to maintain current Tephra semantics (read-your-writes), while providing the knobs necessary to prevent this inducing an infinite loop scenario (the UPSERT/SELECT to same table).  This implementation does this, while only imposing new overhead where it is actually required.

If you would like to introduce alternate isolation levels to Tephra, you are welcome to do so, but that is outside the scope of this issue.

The solution that you are proposing, as far as I understand it, seems to have two problems:

1. When a transaction is started it provides read-your-writes semantics, but as soon as a checkpoint is performed, writes at the current write pointer are no longer visible.  This seems like inconsistent behavior and difficult to reason about since it changes as a side effect of calling the checkpoint operation.
2. As a result, as soon as a checkpoint is performed, a new checkpoint _must_ be performed to see any new writes.  For interleaved write/read operations, this would seem to be a large amount of overhead, where the current solution provides read-your-writes semantics without it.

Phoenix only needs to checkpoint when an UPSERT/SELECT statement is issued against the same table.  At this point it would checkpoint and set the TX_EXCLUDE_CURRENT_WRITE attribute on the Scan operation that it uses for the SELECT portion of that statement.  After this it would not need any special behavior to continue seeing its own writes.  This allows the SELECT portion of the statement to see ALL of the prior writes while just excluding the UPSERTs that are part of the statement.

From @JamesRTaylor's remarks here, it seems like the current implementation would meet their needs.  Again, I do not understand why we would want to build something different.

JamesRTaylor replied:
+1 to get this committed as-is. This meets Phoenix's needs and is the last major JIRA we need to complete the functionality. Keep in mind that Phoenix, not Tephra is providing the SQL semantics. Tephra is broader than this - allowing HBase clients in general to perform transactions - so allow flexibility is a good thing. The reason I like the current implementation is that it doesn't force us to move the write pointer with every statement execution. We only needs this functionality for a statement that reads and writes to the same table. By allowing the client (Phoenix in this case) to decide when this is needed, we gain a performance advantage, as there's a cost to having more and more write pointers that need to be included in the reads for a client.

anew replied:
@ghelmling various approaches have been proposed in this Jira without having the full context of what needs to be achieved. When I suggested that the scan operation could have an option to exclude the current write pointer, that was with the assumption that we may not need checkpointing (or nested transactions - as you had initially named it) for this use case, and I was trying to propose a less invasive approach. That turned out to be not working because it has to see writes made in the same transaction, but before the scan. So I thought we discarded that proposal.

anew replied:
@ghelmling Regarding your other comment. You say I am proposing that 
{quote}
When a transaction is started it provides read-your-writes semantics, but as soon as a checkpoint is performed, writes at the current write pointer are no longer visible.
{quote}
I did not propose that. On the contrary I am proposing that the visibility within a transaction should not change over its lifetime. I am proposing that we introduce a visibility level in transactions, which is either include-own-writes or exclude-own-writes. That becomes a property of the transaction and does not change over its life time. When phoenix executes a sequence of statements in a transaction, it can start a exclude-own-writes transaction. After each statement, it checkpoints the transaction, thus making the writes of all previous statement visible to the subsequent statement. But for the entire life span, the transaction never reads it current write version. 

With that, the visibility is encapsulated inside the Transaction class, similar to how we encapsulate exclusion of invalid or in-progress transactions. If we should ever introduce a "dirty-read" mode, it would also be encapsulated in the Transaction through its isVisible() method. Or would you want to add another boolean flag named includeInProgressWrites? 

The advantage of relying solely on the Transaction for visibility is that existing datasets remain interoperable with Transaction Checkpoints: because the dataset (or tx-aware) itself does not need to implement any new logic for including or excluding the write version. 

With your approach, all existing tx-aware classes have to implement a new feature, namely, the option to exclude the write version. Otherwise they cannot be used in a check-pointed transaction. Encapsulation normally protect us from percolating changes into every related class. This is common practice in modern software engineering, and I am suggesting to keep it that way.

anew replied:
@JamesRTaylor I agree that this meets Phoenix's needs, and as long as the only tx-aware being used is Tephra's transaction-aware HTable, that may be ok. But keep in mind also that Tephra is not only for HBase, for example, CDAP has a whole bunch of tx-aware that are not HBase. 

I am not sure why check-pointing the transaction after every statement would be a problem, if you think that it is too expensive, because the TransactionManager needs to issue a new write pointer, then you may be right. 

Would the folloing compromise work: Instead of requiring a checkpoint after a every statement, we can add a method setVisibility() to the Transaction. That way, you can change the visibility after every statement. And at the same type we keep the encapsulation and don't percolate the change into every tx-aware.

anew commented on tephra-core/src/main/java/co/cask/tephra/TransactionContext.java:
```diff
@@ -83,11 +100,53 @@ public void finish() throws TransactionFailureException {
+      currentTx = txClient.checkpoint(currentTx);
+      // update the current transaction with all TransactionAwares
+      for (TransactionAware txAware : txAwares) {
+        txAware.startTx(currentTx);
```
I don't think it is wise to call startTx on all transactionAwares. startTx is documented to be called at the start of a new transaction. A transaction-aware would therefore clear its state to make sure no change sets or anything else is left from the previous transaction. But I believe in this case we only want the transaction-aware to update its write version to currentTx, but remember all change sets made so far under this transaction.

anew replied:
would it make sense to get the change set from all tx-awares here? If we do that, then tx-awares don't even need to know whether they are part of a check-pointed transaction or not.

ghelmling replied:
The TransactionAwareHTables in Tephra do not behave this way.  But if this is a concern, we can introduce a new method to update the current transaction.

anew commented on tephra-hbase-compat-0.96/src/main/java/co/cask/tephra/hbase96/coprocessor/TransactionProcessor.java:
```diff
@@ -149,10 +149,13 @@ public void preGetOp(ObserverContext<RegionCoprocessorEnvironment> e, Get get, L
     throws IOException {
     Transaction tx = getFromOperation(get);
     if (tx != null) {
+      boolean excludeCurrentWritePtr = isAttributeSet(get, TxConstants.TX_EXCLUDE_CURRENT_WRITE);
```
See my other comments. I think it is better if data operations are completely unaware of whether the transaction excludes the current write version or not. They should simply rely on tx.isVisible() to determine whether a version is visible. That limits the change to the Transaction class itself, and minimizes the change required in every transaction-aware, to be compatible with checkpoints.

ghelmling replied:
This simply supports the ability for clients to exclude the current write pointer from reads on a per-request basis.  I don't see where the TransactionAware needs to know whether or not a checkpoint has been performed.

abaranau replied:
@anew I don't follow your comment. Whether exclude or include current writepointer - is an attribute of an operation/request, not whole transaction. I think things are good here.

anew replied:
@abaranau other visibility constraints - such as excluding uncommitted write versions  - are also attributes of the transaction, not the individual request. I am simply suggesting to follow the existing model. Code is more maintainable if we do similar things in similar ways.

anew commented:
@ghelmling please see my comments, I have some doubts about the API changes (as I had previously made similar comments in the Jira). I hate to make these comments after all the code is written, but I really think we should attempt to keep the API compatible with previous versions, and minimize the changes required for existing transaction-awares to be compatible with checkpoints.

ghelmling commented:
Updated with rename of the Transaction methods:

* getWritePointer() -> getTransactionId()
* getCurrentWritePointer() -> getWritePointer()

ghelmling commented:
Added a new TransactionAware.updateTx() method to clearly differentiate updating the current transaction vs starting a new transaction (and resetting any transaction state).

ghelmling commented:
Added rename of thrift struct fields for transactionId and writePointer, missed on first rename pass.

anew commented:
Regarding backward compatibility. The new method updateTx() in the TransactionAware interface will break backward compatibility. All existing tx-aware implementation will have to be updated to include the new method. I would like to start thinking more about backward-compatibility. One possible approach is:

- add a new interface CheckpointAware extends TransactionAware. This new interface requires the new method updateTx(). 
- add a new class CheckpointingTransactionContext. It requires that all participating TransactionAwares are actually CheckpointAware. 

This means that existing TransactionAware do not break; existing uses of TransactionContext keep working, and Phoenix can use the new CheckpointingTransactionContext with a CheckpointAware HTable. 

Does that make sense? Or did I overlook something?

anew commented:
Re Thrift changes: This will require that the TransactionClient and the TransactionManager are updated at the same time - because of the protocol change, existing clients will not be able to call the new Manager and vice versa. Is there any way we can make this interoperable? The reason for asking is that in CDAP we want to start supporting rolling upgrades. That would not be possible if Tephra changes its wire protocol. If this is too much to ask, we may be able to get away with an incompatible change for now, but we need figure out soon how to address that.

ghelmling commented:
Re Thrift changes: This will require that the TransactionClient and the TransactionManager are 
    updated at the same time - because of the protocol change, existing clients will not be able to call the
    new Manager and vice versa. Is there any way we can make this interoperable? The reason for asking 
    is that in CDAP we want to start supporting rolling upgrades. That would not be possible if Tephra 
    changes its wire protocol. If this is too much to ask, we may be able to get away with an incompatible 
    change for now, but we need figure out soon how to address that.

This is false.  Thrift binary encodings only use the field number not the field name for fields in a struct.  Fields can be renamed without breaking compatibility.

anew commented:
That is good news then, I was not aware of it. Does the addition of a method (checkpoint) have an impact of compatibility? Can we please make sure to test that?

ghelmling commented:
Would the folloing compromise work: Instead of requiring a checkpoint after a every statement, we can
    add a method setVisibility() to the Transaction. That way, you can change the visibility after every 
    statement. And at the same type we keep the encapsulation and don't percolate the change into every
    tx-aware.

So you agree that we should not require clients to perform a new checkpoint every time they wish to read new writes, and that we should use a flag instead?

You just think we should pass the flag to setVisibility() instead of isVisible()?  That's fine.

ghelmling commented:
Regarding backward compatibility. The new method updateTx() in the TransactionAware interface
    will break backward compatibility. All existing tx-aware implementation will have to be updated to 
    include the new method. I would like to start thinking more about backward-compatibility. One 
    possible approach is:

    add a new interface CheckpointAware extends TransactionAware. This new interface requires the 
    new method updateTx().

    add a new class CheckpointingTransactionContext. It requires that all participating 
    TransactionAwares are actually CheckpointAware.

There are already backwards incompatible changes here due to the separation of transactionId and writePointer in the Transaction class.

In addition, modeling features on an inheritance basis generally does not scale well.  What happens when we add a new feature?  Do we add a new interface and context for that?

If we want to avoid breaking implementations of TransactionAware, we should add a BaseTransactionAware class that other implementations can extend in order to inherit default methods.  But I don't think that is required for this change.

anew commented:
1) looks like we agree on a setVisibility() for the transaction. 
2) Is the separation of transaction id and write pointer really incompatible? For transactions that do no check pointing (that is, all existing code), getWritePointer() is exactly what it used to be, right? So I think in this case, we are nicely compatible: existing code does not break, but code that wants to use new functionality needs to use new interfaces.

ghelmling commented:
but code that wants to use new functionality needs to use new interfaces.

As I mentioned, I don't think modeling each new feature in an interface scales very well from the Tephra perspective.  Is there a problem with just updating the TransactionAware implementations when upgrading to this version of Tephra?  Implementing updateTx() seems pretty straightforward.

anew commented:
It requires recompiling and redeploying all apps at the time of the upgrade. Which is something we would like to avoid when possible.

ghelmling commented:
It requires recompiling and redeploying all apps at the time of the upgrade. Which is something we 
    would like to avoid when possible.

Can you explain why this would be necessary?  The JLS Binary Compatibility section states that adding methods to a superclass does not break binary compatibility (section 13.4.12 and 13.5.3):
https://docs.oracle.com/javase/specs/jls/se7/html/jls-13.html

Regardless, any custom dataset would be extending AbstractDataset, which implements TransactionAware.  AbstractDataset would implement updateTx(Transaction) and is part of the platform code, so the new version would be loaded after upgrade.  Any custom dataset should continue to work.

Even if application code were to implement TransactionAware directly, this would not break binary compatibility according to section 13.5.3 of the Binary Compatibility chapter, even if the class would no longer compile without code changes.

ghelmling commented:
Rebased against develop to incorporate TEPHRA-92 changes.  Removed use of the TxConstants.TX_EXCLUDE_CURRENT_WRITE attribute.  Added VisibilityLevel enum and setVisibility(VisibilityLevel) to Transaction.

anew commented:
I would like to ask @abaranau to finish this code review, because I will be away for the next 4 days. Thank you!

nitinmotgi commented:
I reviewed it and It looks good to me. LGTM :+1: Please merge. @anew if there are any issues we will revisit it again.
